raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1859,15 +1819,17 @@ private void appendBatch(
                 offsetAndEpoch.offset + 1, Integer.MAX_VALUE);
 
             future.whenComplete((commitTimeMs, exception) -> {
-                int numRecords = batch.records.size();
-                if (exception != null) {
-                    logger.debug("Failed to commit {} records at {}", numRecords, offsetAndEpoch, exception);
-                } else {
-                    long elapsedTime = Math.max(0, commitTimeMs - appendTimeMs);
-                    double elapsedTimePerRecord = (double) elapsedTime / numRecords;
-                    kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, appendTimeMs);
-                    logger.debug("Completed commit of {} records at {}", numRecords, offsetAndEpoch);
-                    maybeFireHandleCommit(batch.baseOffset, epoch, batch.records);
+                if (batch.records.isPresent()) {
```

Comment:
It would be nice if we didn't lose the exception for the leader change message. Perhaps one way we can do this is to add `numRecords` as a separate field in `CompletedBatch` so that we always have it available. Then we could change this to something like this:
```java
future.whenComplete((commitTimeMs, exception) -> {
  if (exception != null) {
    logger.debug("Failed to commit {} records at {}", batch.numRecords, offsetAndEpoch, exception);
  } else {
    long elapsedTime = Math.max(0, commitTimeMs - appendTimeMs);
    double elapsedTimePerRecord = (double) elapsedTime / batch.numRecords;
    kafkaRaftMetrics.updateCommitLatency(elapsedTimePerRecord, appendTimeMs);
    logger.debug("Completed commit of {} records at {}", batch.numRecords, offsetAndEpoch);
    batch.records.ifPresent(records -> maybeFireHandleCommit(batch.baseOffset, epoch, records));
  }
}
```